subprojects/language-native/src/main/java/org/gradle/language/plugins/NativeBasePlugin.java

```diff
@@ -429,4 +468,55 @@ public void execute(MultipleCandidatesDetails<Linkage> details) {
             }
         }
     }
+
+    static class HeaderDirectoryCompatibilityRule implements AttributeCompatibilityRule<String> {
+        @Override
+        public void execute(CompatibilityCheckDetails<String> details) {
+            // This allows arbitrary file-based dependencies that are added without attributes
+            // to be recognized as compatible for C++ headers if they're directories
+            if (C_PLUS_PLUS_API_DIRECTORY.equals(details.getConsumerValue())
+                    && DIRECTORY.equals(details.getProducerValue())) {
+                details.compatible();
+            }
+        }
+    }
+
+    private static class UnzipTransform extends ArtifactTransform {
+        @Inject
+        UnzipTransform() { }
```

Comment:
Unnecessary?

Reply:
Necessary.  A constructor with @Inject seems to be required for artifact transforms.

Reply:
Ah, as it turns out it was only necessary because it was a private inner class - for some reason this didn't occur to me before, but seems obvious now - once we moved it to its own top-level class, I could remove this.